### PR TITLE
Fix runtime behaviour of PEP 696

### PR DESCRIPTION
**Pull request by NCPlayz**
In [PEP 696](https://peps.python.org/pep-0696/) there are two failure modes specified:

```py
DefaultStrT = TypeVar("DefaultStrT", default=str)
DefaultIntT = TypeVar("DefaultIntT", default=int)
DefaultBoolT = TypeVar("DefaultBoolT", default=bool)
T = TypeVar("T")
T2 = TypeVar("T2")

 # 1. Invalid: non-default TypeVars cannot follow ones with defaults
class NonDefaultFollowsDefault(Generic[DefaultStrT, T]): ...

class AllTheDefaults(Generic[T1, T2, DefaultStrT, DefaultIntT, DefaultBoolT]): ...

# 2. Invalid: expected 2 arguments to AllTheDefaults
AllTheDefaults[int]
```

This PR aims to implement the runtime behaviour for this PEP that was missed out in the initial support PR #77. 
It also brings support for this to Python 3.11+.

One note I had was that the current error will say `Too few parameters for AllTheDefaults; actual 1, expected 5`.
For TypeVarLikes with defaults they aren't actually required, so is it worth computing the minimum required then saying `expected at least 2` for this case?

Let me know if you have any suggestions!

### CODE REVIEW

File: src/test_typing_extensions.py [hunk 1]
```diff
@@ -5605,6 +5605,23 @@ def test_typevartuple(self):
         class A(Generic[Unpack[Ts]]): ...
         Alias = Optional[Unpack[Ts]]
 
+    def test_erroneous_generic(self):
+        DefaultStrT = TypeVar('DefaultStrT', default=str)
```
Comment (JelleZijlstra) [inline src/test_typing_extensions.py:5609]: A bit weird that this one uses plain TypeVar while the next test uses `typing_extensions.TypeVar`. They are the same (we do `from typing_extensions import TypeVar`), so we should stick to just one variation. I think it is clearer to use `typing_extensions.TypeVar` but don't feel strongly.

(No need to change this in other tests.)

File: src/typing_extensions.py [hunk 2]
```diff
@@ -2378,9 +2336,139 @@ def wrapper(*args, **kwargs):
...
+                    for collected in _collect_parameters([x]):
+                        if collected not in parameters:
+                            parameters.append(collected)
+            elif hasattr(t, '__typing_subst__'):
+                if t not in parameters:
+                    if getattr(t, '__default__', None) is not None:
+                        if not default_encountered:
+                            default_encountered = True
+                    elif default_encountered:
+                        raise TypeError('expected TypeVar with default type, found'
```
Comment (JelleZijlstra) [inline src/typing_extensions.py:2460]: I think this error message could be clearer. Also, TypeVar seems wrong as it could be a ParamSpec or TypeVarTuple instead.

For inspiration:
```
% python -c 'def f(x=1, y): ...'
  File "<string>", line 1
    def f(x=1, y): ...
               ^
SyntaxError: parameter without a default follows parameter with a default
```

So maybe "type parameter {t!r} without a default follows type parameter with a default".

File: src/typing_extensions.py [hunk 3]
```diff
@@ -2378,9 +2336,139 @@ def wrapper(*args, **kwargs):
...
+                # deal with TypeVarLike defaults
+                # required TypeVarLikes cannot appear after a defaulted one.
+                if alen < elen:
+                    # since we validate TypeVarLike default in _collect_type_vars
+                    # or _collect_parameters we can safely check parameters[alen]
+                    if getattr(parameters[alen], '__default__', None) is not None:
+                        return
+
+            raise TypeError(f"Too {'many' if alen > elen else 'few'} parameters"
+                            f" for {cls}; actual {alen}, expected {elen}")
```
Comment (JelleZijlstra) [inline src/typing_extensions.py:2366]: Agree the "expected {elen}" part here is problematic, as the precise number of type parameters accepted isn't necessarily fixed. I would look at the error messages from function calls in Python for how to make this better.

**Review by JelleZijlstra — COMMENTED** [review]
Thanks! cc @Gobot1234

File: src/typing_extensions.py [hunk 4]
```diff
@@ -145,27 +145,6 @@ def __repr__(self):
...
-            raise TypeError(f"{cls} is not a generic class")
-        elen = len(cls.__parameters__)
-    alen = len(parameters)
-    if alen != elen:
-        if hasattr(cls, "__parameters__"):
-            parameters = [p for p in cls.__parameters__ if not _is_unpack(p)]
-            num_tv_tuples = sum(isinstance(p, TypeVarTuple) for p in parameters)
-            if (num_tv_tuples > 0) and (alen >= elen - num_tv_tuples):
-                return
-        raise TypeError(f"Too {'many' if alen > elen else 'few'} parameters for {cls};"
```
Comment (NCPlayz) [inline src/typing_extensions.py:165]: Was trying to figure out why the tests failed when I reuse this error and it is because typing_extensions raises with "parameters" whereas Python has always raised with "arguments", and this conditional is placed everywhere in tests:

https://github.com/python/typing_extensions/blob/fda0c15e3b5ed05703420cfb4d0974edb5e39c46/src/test_typing_extensions.py#L5234

In Python 3.12:

https://github.com/python/cpython/blob/cf29a2f25e359790c0c0f6a4ab0ab49ca5ffab5d/Lib/typing.py#L297

In Python 3.10:

https://github.com/python/cpython/blob/dcb16c98be61630369227f0d893f8d9262d25cac/Lib/typing.py#L231

Is this phrase change on purpose and should the tests be fixed accordingly?

File: src/typing_extensions.py (hunk 3 above)
Comment (NCPlayz) [reply src/typing_extensions.py:2366]: Using https://github.com/python/cpython/blob/244567398370bfa62a332676762bb1db395c02fc/Lib/typing.py#L974-L975 as inspiration to change the error message

**Comment by cdce8p** [comment]
What's the status of this one? With basic / experimental support for PEP 696 coming in the next mypy release this is basically necessary to use TypeVar defaults.

File: src/test_typing_extensions.py [hunk 5]
```diff
@@ -5711,9 +5714,7 @@ class Y(Generic[T], NamedTuple):
                 self.assertIsInstance(a, G)
                 self.assertEqual(a.x, 3)
 
-                things = "arguments" if sys.version_info >= (3, 11) else "parameters"
-
-                with self.assertRaisesRegex(TypeError, f'Too many {things}'):
+                with self.assertRaisesRegex(TypeError, f'Too many parameters'):
```
Comment (Gobot1234) [inline src/test_typing_extensions.py:5717]:
```suggestion
                with self.assertRaisesRegex(TypeError, 'Too many parameters'):
```
To fix CI

**Review by Gobot1234 — COMMENTED** [review]
I don't have the necessary internal knowledge of how typing.py works anymore really to be able to review this properly

**Comment by cdce8p** [comment]
> It would be good to get more review (if @Gobot1234 or you could review, I'd appreciate it).

👍🏻  I'll look at it tomorrow.

File: src/typing_extensions.py [hunk 6]
```diff
@@ -2689,9 +2647,154 @@ def wrapper(*args, **kwargs):
 #   counting generic parameters, so that when we subscript a generic,
 #   the runtime doesn't try to substitute the Unpack with the subscripted type.
 if not hasattr(typing, "TypeVarTuple"):
+    def _check_generic(cls, parameters, elen=_marker):
+        """Check correct count for parameters of a generic cls (internal helper).
+        This gives a nice error message in case of count mismatch.
+        """
```
Comment (cdce8p) [inline src/typing_extensions.py:2653]:
```suggestion
        """Check correct count for parameters of a generic cls (internal helper).

        This gives a nice error message in case of count mismatch.
        """
```
Insert an empty line to match the docstring in the `else` case.

File: src/typing_extensions.py [hunk 7]
```diff
@@ -2689,9 +2647,154 @@ def wrapper(*args, **kwargs):
...
+        default_encountered = False
+        for t in types:
+            if (
+                isinstance(t, typevar_types) and
+                t not in tvars and
+                not _is_unpack(t)
+            ):
+                if getattr(t, '__default__', None) is not None:
+                    if not default_encountered:
+                        default_encountered = True
```
Comment (cdce8p) [inline src/typing_extensions.py:2746]:
```suggestion
                    default_encountered = True
```
This can be simplified. The `if` check is unnecessary.

File: src/typing_extensions.py [hunk 8]
```diff
@@ -2689,9 +2647,154 @@ def wrapper(*args, **kwargs):
...
+                # `[T, int]`, or `[int, *Ts]`, etc.
+                for x in t:
+                    for collected in _collect_parameters([x]):
+                        if collected not in parameters:
+                            parameters.append(collected)
+            elif hasattr(t, '__typing_subst__'):
+                if t not in parameters:
+                    if getattr(t, '__default__', None) is not None:
+                        if not default_encountered:
+                            default_encountered = True
```
Comment (cdce8p) [inline src/typing_extensions.py:2784]:
```suggestion
                        default_encountered = True
```

File: src/typing_extensions.py [hunk 9]
```diff
@@ -2689,9 +2647,154 @@ def wrapper(*args, **kwargs):
...
+                isinstance(t, typevar_types) and
+                t not in tvars and
+                not _is_unpack(t)
+            ):
+                if getattr(t, '__default__', None) is not None:
+                    if not default_encountered:
+                        default_encountered = True
+                elif default_encountered:
+                    raise TypeError(f'type parameter {t!r} without a default'
+                                    ' follows type parameter with a default')
```
Comment (cdce8p) [inline src/typing_extensions.py:2749]:
```suggestion
                    raise TypeError(f'Type parameter {t!r} without a default'
                                    ' follows type parameter with a default')
```
I would suggest to capitalize the first word to match the other error messages.

File: src/typing_extensions.py [hunk 10]
```diff
@@ -2689,9 +2647,154 @@ def wrapper(*args, **kwargs):
...
+                        if collected not in parameters:
+                            parameters.append(collected)
+            elif hasattr(t, '__typing_subst__'):
+                if t not in parameters:
+                    if getattr(t, '__default__', None) is not None:
+                        if not default_encountered:
+                            default_encountered = True
+                    elif default_encountered:
+                        raise TypeError(f'type parameter {t!r} without a default'
+                                        ' follows type parameter with a default')
```
Comment (cdce8p) [inline src/typing_extensions.py:2787]:
```suggestion
                        raise TypeError(f'Type parameter {t!r} without a default'
                                        ' follows type parameter with a default')
```

File: src/typing_extensions.py [hunk 11]
```diff
@@ -2689,9 +2647,154 @@ def wrapper(*args, **kwargs):
...
+
+                    num_default_tv = sum(getattr(p, '__default__', None)
+                                         is not None for p in parameters)
+
+                    elen -= num_default_tv
+
+                    expect_val = f"at least {elen}"
+
+            raise TypeError(f"Too {'many' if alen > elen else 'few'} parameters"
+                            f" for {cls}; actual {alen}, expected {expect_val}")
```
Comment (cdce8p) [inline src/typing_extensions.py:2685]:
```suggestion
            raise TypeError(f"Too {'many' if alen > elen else 'few'} arguments"
                            f" for {cls}; actual {alen}, expected {expect_val}")
```
If we unify `parameters` and `arguments`, I believe it makes sense to use the "newer" one. I.e. Python 3.11+ uses `arguments`.

File: src/typing_extensions.py [hunk 12]
```diff
@@ -2689,9 +2647,154 @@ def wrapper(*args, **kwargs):
...
```
Comment (cdce8p) [inline src/typing_extensions.py:2721]:
```suggestion
            raise TypeError(f"Too {'many' if alen > elen else 'few'} arguments"
                            f" for {cls}; actual {alen}, expected {expect_val}")
```

File: src/test_typing_extensions.py [hunk 13]
```diff
@@ -3262,6 +3262,9 @@ def __call__(self, *args: Unpack[Ts]) -> T: ...
         self.assertEqual(MemoizedFunc.__parameters__, (Ts, T, T2))
         self.assertTrue(MemoizedFunc._is_protocol)
 
+        # 3.11+ moves where this exception is thrown
+        # 3.11: TypeVarTuple.__typing_prepare_subst__
+        # 3.12+: typing._typevartuple_prepare_subst
```
Comment (cdce8p) [inline src/test_typing_extensions.py:3267]:
```suggestion
```
Unrelated?

File: src/test_typing_extensions.py [hunk 14]
```diff
@@ -3262,6 +3262,9 @@ def __call__(self, *args: Unpack[Ts]) -> T: ...
...
         things = "arguments" if sys.version_info >= (3, 11) else "parameters"
```
Comment (cdce8p) [inline src/test_typing_extensions.py:3268]: This is no longer necessary if we use `arguments` for all versions.
The `assertRaisesRegex` below should than also be updated to use `arguments` instead.
https://github.com/python/typing_extensions/blob/4fdc09ddb54be26580f68e26443a422c6024364c/src/test_typing_extensions.py#L3272-L3273

File: src/test_typing_extensions.py [hunk 15]
```diff
@@ -5711,9 +5714,7 @@ class Y(Generic[T], NamedTuple):
                 self.assertIsInstance(a, G)
                 self.assertEqual(a.x, 3)
 
-                things = "arguments" if sys.version_info >= (3, 11) else "parameters"
-
-                with self.assertRaisesRegex(TypeError, f'Too many {things}'):
+                with self.assertRaisesRegex(TypeError, 'Too many parameters'):
```
Comment (cdce8p) [inline src/test_typing_extensions.py:5717]:
```suggestion
                with self.assertRaisesRegex(TypeError, 'Too many arguments'):
```

File: src/test_typing_extensions.py (hunk 14 above)
Comment (cdce8p) [inline src/test_typing_extensions.py:3268]:
```suggestion
        things = "arguments"
```

File: src/test_typing_extensions.py [hunk 16]
```diff
@@ -6215,6 +6216,23 @@ def test_typevartuple(self):
...
+
+        with self.assertRaises(TypeError):
+            Test = Generic[DefaultStrT, T]
+
+    def test_need_more_params(self):
+        DefaultStrT = typing_extensions.TypeVar('DefaultStrT', default=str)
+        T = typing_extensions.TypeVar('T')
+        U = typing_extensions.TypeVar('U')
+
+        class A(Generic[T, U, DefaultStrT]): ...
```
Comment (cdce8p) [inline src/test_typing_extensions.py:6231]:
```suggestion
        class A(Generic[T, U, DefaultStrT]): ...
        A[int, bool]
        A[int, bool, str]
```
I would add two good examples here, just so we test them too.

File: src/test_typing_extensions.py [hunk 17]
```diff
@@ -6215,6 +6216,23 @@ def test_typevartuple(self):
...
+
+        with self.assertRaises(TypeError):
```
Comment (cdce8p) [inline src/test_typing_extensions.py:6233]:
```suggestion
        with self.assertRaises(
            TypeError, "Too few arguments for .+; actual 1, expected at least 2"
        ):
```

**Review by cdce8p — COMMENTED** [review]
Left a few comments. Most of them are just some minor style ones.
Overall this looks good to me! Thanks @NCPlayz 👍🏻 

--
@JelleZijlstra I left suggestions everywhere I could. If you like feel free to merge them. Would be awesome to get this released soon.

**Comment by NCPlayz** [comment]
Thanks for the reviews! Fixed a minor bug but everything seems to be passing.

File: src/typing_extensions.py [hunk 18]
```diff
@@ -2689,9 +2647,153 @@ def wrapper(*args, **kwargs):
...
+        """
+        if not elen:
+            raise TypeError(f"{cls} is not a generic class")
+        alen = len(parameters)
+        if alen != elen:
+            expect_val = elen
+            if hasattr(cls, "__parameters__"):
+                parameters = [p for p in cls.__parameters__ if not _is_unpack(p)]
+                num_tv_tuples = sum(isinstance(p, TypeVarTuple) for p in parameters)
+                if (num_tv_tuples > 0) and (alen >= elen - num_tv_tuples):
```
Comment (JelleZijlstra) [inline src/typing_extensions.py:2703]: This condition isn't present on main, why do we need it here?

https://github.com/python/cpython/blob/2339e7cff745271f0e4a919573a347ab2bc1c2e9/Lib/typing.py#L289

File: src/typing_extensions.py [hunk 19]
```diff
@@ -2689,9 +2647,153 @@ def wrapper(*args, **kwargs):
...
+                if t not in parameters:
+                    if getattr(t, '__default__', None) is not None:
+                        default_encountered = True
+                    elif default_encountered:
+                        raise TypeError(f'Type parameter {t!r} without a default'
+                                        ' follows type parameter with a default')
+
+                    parameters.append(t)
+            else:
+                if _should_collect_from_parameters(t):
```
Comment (JelleZijlstra) [inline src/typing_extensions.py:2790]: This is different from the behavior on CPython main: https://github.com/python/cpython/blob/2339e7cff745271f0e4a919573a347ab2bc1c2e9/Lib/typing.py#L283

**Review by JelleZijlstra — COMMENTED** [review]
Thanks. We need to be extra careful in the cases where we monkeypatch CPython, so I left some comments there.

I'll look into adding a test to this repo that runs the test_typing suite with the typing_extensions monkeypatches enabled.

**Comment by JelleZijlstra** [comment]
Ci didn't run

**Comment by JelleZijlstra** [comment]
Ci didn't run

**Comment by JelleZijlstra** [comment]
I may have made CI not run any more somehow, will investigate.

File: src/typing_extensions.py (hunk 19 above)
Comment (JelleZijlstra) [reply src/typing_extensions.py:2791]: Looks like this leads to a failure in the test_typing suite, could you make this mirror the CPython logic more closely?

Comment (NCPlayz) [reply src/typing_extensions.py:2791]: Gonna go do that now :pray:

**Comment by NCPlayz** [comment]
Oh.. just noticed that some changes to unit tests might have broken in one of the merge commits

**Comment by JelleZijlstra** [comment]
What specifically is broken? Feel free to send another PR with any fixes.

File: src/test_typing_extensions.py [hunk 20]
```diff
@@ -5712,7 +5712,6 @@ class Y(Generic[T], NamedTuple):
                 self.assertEqual(a.x, 3)
 
                 things = "arguments" if sys.version_info >= (3, 10) else "parameters"
```
Comment (NCPlayz) [inline src/test_typing_extensions.py:5714]: Seems like these lines got re-added -- but maybe they were breaking for 3.8/3.9?

Comment (JelleZijlstra) [reply src/test_typing_extensions.py:5714]: I think so, yes. I see you commented on this in https://github.com/python/typing_extensions/pull/293#discussion_r1375254093 before; I don't feel too strongly but I think it's best to match CPython's behavior.
